### PR TITLE
Fix possible race on container restart and restart error handling

### PR DESCRIPTION
**Pull request by tonistiigi**
Solution for https://github.com/docker/docker/pull/25150/files#r73000548

Supersedes: #25150

@mlaventure @jhowardmsft @cpuguy83

### CODE REVIEW

File: libcontainerd/container_windows.go [hunk 1]
```diff
@@ -260,14 +260,16 @@ func (ctr *container) waitExit(process *process, isFirstProcessToStart bool) err
 			err := <-waitRestart
 			ctr.restarting = false
 			ctr.client.deleteContainer(ctr.friendlyName)
+			if err == nil {
```
Comment (swernli) [inline libcontainerd/container_windows.go]: Perfect, exactly what I was hoping for!

File: libcontainerd/container_linux.go [hunk 2]
```diff
@@ -190,6 +190,10 @@ func (ctr *container) handleEvent(e *containerd.Event) error {
 					ctr.client.lock(ctr.containerID)
 					defer ctr.client.unlock(ctr.containerID)
 					ctr.restarting = false
+					if err == nil {
+						err = ctr.start()
+						logrus.Errorf("libcontainerd: error restarting %v", err)
```
Comment (cpuguy83) [inline libcontainerd/container_linux.go]: should this be wrapped in an `if err != nil`?

File: libcontainerd/container_windows.go [hunk 3]
```diff
@@ -260,14 +260,16 @@ func (ctr *container) waitExit(process *process, isFirstProcessToStart bool) err
...
+				err = ctr.client.Create(ctr.containerID, ctr.ociSpec, ctr.options...)
+				logrus.Errorf("libcontainerd: error restarting %v", err)
```
Comment (cpuguy83) [inline libcontainerd/container_windows.go]: likewise

File: libcontainerd/container_linux.go (hunk 2 above)
Comment (tonistiigi) [inline libcontainerd/container_linux.go]: Fixed. Thanks!

**Comment by mlaventure** [comment]
LGTM.

**Comment by swernli** [comment]
LGTM

**Comment by cpuguy83** [comment]
LGTM

**Pull request merged**